### PR TITLE
docker: move host path for hosts file mount to alloc dir

### LINKED ISSUE #10809: generated /etc/hosts not shared between tasks
### Nomad version
Nomad v1.1.2 (60638a086ef9630e2a9ba1e237e8426192a44244)

### Operating system and Environment details
AWS Linux 2
Docker version:
Client:
 Version:           19.03.6-ce
 API version:       1.40
 Go version:        go1.13.4
 Git commit:        369ce74
 Built:             Fri May 29 04:01:26 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.6-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.4
  Git commit:       369ce74
  Built:            Fri May 29 04:01:57 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.2
  GitCommit:        ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

### Issue

In the latest Nomad release v1.1.2 the [changelog](https://github.com/hashicorp/nomad/blob/main/CHANGELOG.md#112-june-22-2021) mentions the improvement from [GH-10766](https://github.com/hashicorp/nomad/pull/10766):
"docker: Tasks using network.mode = "bridge" that don't set their network_mode will receive a /etc/hosts file that includes the pause container's hostname and any extra_hosts."

I found that this is actually a breaking change from previous behavior in Nomad v1.1.1 when the `network_mode` is not set.

### Reproduction steps

The scenario that I found this to be an issue in is where:
- no `network_mode` is defined for two tasks in a Task Group using a network namespace
- in one task, task A, the `/etc/hosts` file is updated at runtime from the docker entrypoint
- the other task, task B, uses the updated entries in the `/etc/hosts` file to resolve service-to-service mesh routing

In Nomad version <= v1.1.1 the `/etc/hosts` file is updated in both containers but with this new change in v1.1.2 the hosts file is overwritten in Task B (or is not shared anymore?) which breaks out current network setup for routing between containers.

### Workaround
To fix this, we are removing the manual updating of `/etc/hosts` from Task A and instead using the `extra_hosts` feature with the docker driver so both tasks are updated (this is the way we should have done it from the get go but this was a forcing function). We're currently making this change to restore networking in our environment but it came to us as a surprise when we updated and all new deployments could no longer talk to other services.

#### Expected Result

The CHANGELOG.md reflects the breaking change or a change is implemented that adjusts this behavior to continue shared R/W access of `/etc/hosts` between the pause, task A, and task B container

#### Actual Result

[surprised_pikachu.jpg ](https://i.kym-cdn.com/photos/images/newsfeed/001/431/201/40f.png) upon updating to the latest Nomad when the `/etc/hosts` no longer appears to be shared between network namespaced containers

### PR DESCRIPTION
**Pull request by tgross**
Fixes #10809 

In Nomad 1.1.1 we generate a hosts file based on the Nomad-owned network
namespace, rather than using the default hosts file from the pause
container. This hosts file should be shared between tasks in the same
allocation so that tasks can update the file and have the results propagated
between tasks.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,8 @@
 ## 1.1.3 (Unreleased)
 
 BUG FIXES:
-csi: fixed a CLI panic when formatting `volume status` with `-verbose` flag [[GH-10818](https://github.com/hashicorp/nomad/issues/10818)]
+* csi: Fixed a CLI panic when formatting `volume status` with `-verbose` flag [[GH-10818](https://github.com/hashicorp/nomad/issues/10818)]
+* docker: Moved the generated `/etc/hosts` file's mount source to the allocation directory so that it can be shared between tasks of an allocation. [[GH-10823](https://github.com/hashicorp/nomad/issues/10823)]
```
Comment (tgross) [inline CHANGELOG.md:5]: I'd opened this PR before https://github.com/hashicorp/nomad/pull/10825... should I wait to merge and then rebase with the new changelog files?

**Comment by shoenig** [comment]
LGTM; is this something we should call out in the upgrade notes? (as unlikely as anyone is doing so, currently tasks could modify this file independently and now they can't, right?)

**Comment by tgross** [comment]
> LGTM; is this something we should call out in the upgrade notes? (as unlikely as anyone is doing so, currently tasks could modify this file independently and now they can't, right?)

Seems worth calling out.

File: CHANGELOG.md (hunk 1 above)
Comment (tgross) [reply CHANGELOG.md]: Done

**Review by shoenig — APPROVED** [review]

**Pull request merged**